lua/packer/clean.lua

```diff
@@ -9,44 +9,71 @@ local async = a.sync
 
 local config
 
+local PLUGIN_OPTIONAL_LIST = 1
+local PLUGIN_START_LIST = 2
+
 local function is_dirty(plugin, typ)
-  return plugin.disable or (plugin.opt and typ == 2) or (not plugin.opt and typ == 1)
+  return plugin.disable or (plugin.opt and typ == PLUGIN_START_LIST) or (not plugin.opt and typ == PLUGIN_OPTIONAL_LIST)
 end
 
 -- Find and remove any plugins not currently configured for use
 local clean_plugins = function(_, plugins, results)
   return async(function()
     results = results or {}
     results.removals = results.removals or {}
-    local opt_plugins, start_plugins = plugin_utils.list_installed_plugins()
-    local dirty_plugins = {}
-    local aliases = {}
-    for _, plugin in pairs(plugins) do
-      if plugin.as and not plugin.disable then aliases[plugin.as] = true end
+
+    local function map_install_folder(install_paths)
+      local install_folders = {}
+
+      for install_path, _ in pairs(install_paths) do
+        local split_path = vim.split(install_path, '/', true)
+        install_folders[split_path[#split_path]] = true
+      end
+
+      return install_folders
     end
 
-    for typ, plugin_list in ipairs({opt_plugins, start_plugins}) do
-      for plugin_path, _ in pairs(plugin_list) do
-        local plugin_name = vim.fn.fnamemodify(plugin_path, ":t")
-        local plugin_data = plugins[plugin_name]
-        if (plugin_data == nil and not aliases[plugin_name])
-          or (plugin_data and is_dirty(plugin_data, typ)) then
-          dirty_plugins[plugin_path] = plugin_name
-        end
+    local function map_names_to_paths(plugin_names)
+      local paths = {}
+
+      for _, plugin_name in ipairs(plugin_names) do
+        paths[#paths+1] = plugins[plugin_name].install_path
+      end
+
+      return paths
+    end
+
+    local opt_plugins, start_plugins = plugin_utils.list_installed_plugins()
+    local dirty_plugins = plugin_utils.find_missing_plugins(plugins, opt_plugins, start_plugins)
+
+    opt_plugins = map_install_folder(opt_plugins)
+    start_plugins = map_install_folder(start_plugins)
+
+    for _, plugin_config in pairs(plugins) do
+      local plugin_name = plugin_config.short_name
+      local plugin_source = (opt_plugins[plugin_name] and PLUGIN_OPTIONAL_LIST) or
```

Comment:
I'm going to add something here that removes detected plugins from the list, so that anything left over must not have been in the plugin spec and should be removed.